src/OpenCL.cpp

```diff
@@ -507,107 +449,119 @@ void OpenCL_Network::forward(const std::vector<net_t>& input,
     queue.finish();
 }
 
-void OpenCL_Network::convolve(int filter_size, int channels, int outputs,
-                              cl::Buffer& bufferInput,
-                              cl::Buffer& bufferOutput,
-                              cl::Buffer& bufferMerge,
-                              weight_slice_t weights) {
-    // fixed for 19x19
-    constexpr int width = 19;
-    constexpr int height = 19;
-    constexpr int boardsize = width * height;
-
-    cl::Kernel * m_convolve_kernel = nullptr;
-    if (filter_size == 3) {
-        m_convolve_kernel = &opencl_thread_data.m_convolve3_kernel;
-    } else {
-        assert(filter_size == 1);
-        m_convolve_kernel = &opencl_thread_data.m_convolve1_kernel;
-    }
+void OpenCL_Network::convolve3(int channels, int outputs,
+                              cl::Buffer& bufferInOut,
+                              cl::Buffer& bufferV,
+                              cl::Buffer& bufferM,
+                              weight_slice_t weights,
+                              cl::Buffer* bufferResidual,
+                              weight_slice_t* bn_weights) {
+
+    cl::Kernel in_transform_kernel = opencl_thread_data.m_in_transform_kernel;
+    cl::Kernel sgemm_kernel = opencl_thread_data.m_sgemm_kernel;
+    cl::Kernel out_transform_kernel = opencl_thread_data.m_out_transform_kernel;
+    cl::Kernel out_transform_bn_kernel = opencl_thread_data.m_out_transform_bn_kernel;
+
+    auto mwg = opencl.m_sgemm_mwg;
+    auto nwg = opencl.m_sgemm_nwg;
+    auto kwg = opencl.m_sgemm_kwg;
+    auto vwm = opencl.m_sgemm_vwm;
+    auto vwn = opencl.m_sgemm_vwn;
+    auto mdimc = opencl.m_sgemm_mdimc;
+    auto ndimc = opencl.m_sgemm_ndimc;
+    auto wavefront_size = opencl.m_wavefront_size;
+
+    assert(mwg != 0);
+    assert(nwg != 0);
+    assert(kwg != 0);
+    assert(mdimc != 0);
+    assert(ndimc != 0);
+    assert(vwm != 0);
+    assert(vwn != 0);
+    assert(wavefront_size != 0);
+
+    constexpr int tiles = (19 + 1) * (19 + 1) / 4;
+
+    auto wgs = lcm(tiles, wavefront_size);
+
+    int m_ceil = (int)lcm(lcm(outputs, mwg), vwm);
+    int n_ceil = (int)lcm(lcm(tiles, nwg), vwn);
+    int k_ceil = (int)lcm(lcm(channels, kwg), vwm);
 
-    // Input channel grouping in multiples of 8
-    int channelGroup = 8;
-    int channelShift = 3;
+    cl::CommandQueue & queue = opencl_thread_data.m_commandqueue;
 
-    // Input layer is not a multiple of 8
-    if (channels % 8 != 0) {
-        assert(channels % 2 == 0);
-        channelGroup = 2;
-        channelShift = 1;
+    try {
+        in_transform_kernel.setArg(0, bufferInOut);
+        in_transform_kernel.setArg(1, bufferV);
+        in_transform_kernel.setArg(2, channels);
+        in_transform_kernel.setArg(3, k_ceil);
+        in_transform_kernel.setArg(4, n_ceil);
+
+        queue.enqueueNDRangeKernel(in_transform_kernel, cl::NullRange,
+                                   cl::NDRange(wgs, channels));
```

Comment:
I think this introduces a problem that if channels is not a power of 2, you require OpenCL 2.0. (The current code works with e.g. 192 or 384)

In general for squeezing performance out of OpenCL you're going to want to control the workgroup size and 2d blocking yourself.

Reply:
This works with non-power of two arguments for me and I don't think I'm using OpenCL 2.0.

Reply:
Hmm, I guess it's OK because you're not specifying a local range, so it will find a divisor of your global one.